hoodie-spark/src/main/scala/com/uber/hoodie/HoodieStreamingSink.scala

```diff
@@ -0,0 +1,91 @@
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *
+ */
+package com.uber.hoodie
+
+import org.apache.spark.sql.{DataFrame, SQLContext, SaveMode}
+import org.apache.spark.sql.execution.streaming.Sink
+import org.apache.spark.sql.streaming.OutputMode
+import org.apache.log4j.LogManager
+
+import scala.util.{Failure, Success, Try}
+
+class HoodieStreamingSink(sqlContext: SQLContext,
+                          options: Map[String, String],
+                          partitionColumns: Seq[String],
+                          outputMode: OutputMode)
+  extends Sink
+    with Serializable {
+  @volatile private var latestBatchId = -1L
+
+  private val log = LogManager.getLogger(classOf[HoodieStreamingSink])
+
+  private val retryCnt = options(DataSourceWriteOptions.STREAMING_RETRY_CNT_OPT_KEY).toInt
+  private val retryIntervalMs = options(DataSourceWriteOptions.STREAMING_RETRY_INTERVAL_MS_OPT_KEY).toLong
+  private val failOnDataLoss = options(DataSourceWriteOptions.STREAMING_FAILURE_ON_DATA_LOSS_OPT_KEY).toBoolean
+
+  private val mode =
+    if (outputMode == OutputMode.Append()) {
+      SaveMode.Append
+    } else {
+      SaveMode.Overwrite
+    }
+
+  override def addBatch(batchId: Long, data: DataFrame): Unit = {
+    retry(retryCnt, retryIntervalMs)(
+      Try(
+        HoodieSparkSqlWriter.write(
+          sqlContext,
+          mode,
+          options,
+          data)
+      )
+    ) match {
+      case Success((true, _)) =>
+        log.debug(s"Micro batch id=$batchId succeeded")
+      case Failure(e) =>
+        if (failOnDataLoss) {
+          log.error(s"Micro batch id=$batchId threw following expections after $retryCnt retries," +
+            s"aborting streaming to avoid data loss: ", e)
+          // this is to force exit upon error
+          // spark sometimes hangs upon exceptions
+          // will require redeployment / cluster mode to resume
+          System.exit(1)
+        }
+      case Success((false, commitOps)) =>
+        if (failOnDataLoss) {
+          commitOps.foreach{commit => log.error(s"Micro batch failed at commit: $commit")}
+          log.error(s"Micro batch id=$batchId failed after $retryCnt retries due to write or sync failures," +
+            s"aborting streaming to avoid data loss.")
+          System.exit(1)
+        }
+    }
+  }
+
+  override def toString: String = s"HoodieStreamingSink[${options("path")}]"
+
+  @annotation.tailrec
+  private def retry[T](n: Int, waitInMillis: Long)(fn: => Try[T]): Try[T] = {
+    fn match {
+      case x: util.Success[T] => x
+      case _ if n > 1 =>
+        Thread.sleep(waitInMillis)
```

Comment:
Does spark immediately retry?

Reply:
My understanding is that (feel free to correct me if I am wrong), spark only retry on the task level in executors and the retries are immediate.

However this retry is at microbatch level and done in the driver, it is up to user to config `waitInMillis` to be 0 to have immediate retry

Are you saying we should enforce immediate retry here at microbatch level to match spark behavior?

Reply:
No what you have makes sense.. I was just confirming my understanding.. This is good.

Reply:
👍